### PR TITLE
Fix `lll_reduce` for slab generation

### PR DESCRIPTION
**Pull request by jinlhr542**
By making the termination identification process for interface generation more robust.

1. fixing problem for the lll_reduce process when making slabs, doing mapping before updating the structure
2. allow to set ftol of the termination distances for hierarchical cluster so that some non-identical terminations close to each other can be identified
3. allow to add index for terminations so that terminations with the same space group can be distinguished
4. allow to not filter symmetic film and substrate slabs when finding terminations

### CODE REVIEW

**Comment by jinlhr542** [comment]
@DanielYang59 Sorry, it seems that my changes have conflicts with the recent updates, I remake another pull request using the recent version

**Comment by DanielYang59** [comment]
I understand merge conflicts could be confusing but don't worry. Merge conflicts are common things that would happen when you and another person try to make changes to the same line of code. When such things happen, you just need to choose whatever [conflict resolver that works best for you](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-using-the-command-line) (GUI/CLI).

In most cases, reopen a PR is not necessary, perhaps you could try to resolve them next time when such things happen.

I would ping @janosh for review as I'm not a maintainer and therefore don't have write access to the repo. Can I ask for your patience as he has been quite busy recently? Thank you for your effort towards correcting and improving the code.

**Comment by jinlhr542** [comment]
Many thanks for the suggestion... I have another issue here. I found that installing the package by `pip install pymatgen` by `pip install .` at the forked folder have different effects. For me, I cannot successfully import the modules like pymatgen.core, pymatgen.analysis etc... by installing the recently released version on 2024.6.10. May I have some comments on this issue?
image

**Comment by DanielYang59** [comment]
> I found that installing the package and by `pip install pymatgen` by `pip install .` at the forked folder have different effects. 

That's exactly right observation. By `pip install pymatgen`, you are installing `pymatgen` from PyPI, and by running `pip install .`, you are instead installing from a local project, refer to [this](https://pip.pypa.io/en/stable/cli/pip_install/) for more details if you are interested.

Meanwhile, if you were trying to contribute to the code, you might want to use `pip install --editable .` or `pip install -e .` for short in [editable mode](https://pip.pypa.io/en/stable/cli/pip_install/), such that your changes to the code could be tested in real time. 

> For me, I cannot successfully import the modules like pymatgen.core, pymatgen.analysis etc... by installing the recently released version on 2024.6.10 I forked to local folder. 

 I didn't manage to recreate this issue from my side. Perhaps you should double check if you have a file called `pymatgen.py` locally which might have shadowed the real `pymaygen` you are trying to import?

**Comment by jinlhr542** [comment]
Thank you for the explanation but I found the that the reason might be the package is not fully installed, this is the pymatgen folder installed using conda environment... As you can see most of the modules were not successfully installed
image

**Comment by DanielYang59** [comment]
It's weird, but it's hard to give any advice without seeing the log. Can you show me the log from `pip install -e .`? Meanwhile I guess it's more readable for me, and easier for you to just copy and paste the terminal outputs. You could also use [markdown grammars](https://www.markdownguide.org/cheat-sheet/) like triple \``` in both sides of the code to highlight it,or [add a collapsible section](https://gist.github.com/pierrejoubert73/902cc94d79424356a8d20be2b382e1ab) if the log is too long.

**Comment by jinlhr542** [comment]
```(3.11) jason@YaoshuXiedeMacBook-Pro pymatgen % pip uninstall pymatgen
Found existing installation: pymatgen 2024.6.10
Uninstalling pymatgen-2024.6.10:
  Would remove:
    /opt/anaconda3/envs/3.11/bin/feff_plot_cross_section
    /opt/anaconda3/envs/3.11/bin/feff_plot_dos
    /opt/anaconda3/envs/3.11/bin/get_environment
    /opt/anaconda3/envs/3.11/bin/pmg
    /opt/anaconda3/envs/3.11/lib/python3.11/site-packages/__editable__.pymatgen-2024.6.10.pth
    /opt/anaconda3/envs/3.11/lib/python3.11/site-packages/pymatgen-2024.6.10.dist-info/*
Proceed (Y/n)? y
  Successfully uninstalled pymatgen-2024.6.10
(3.11) jason@YaoshuXiedeMacBook-Pro pymatgen % pip install -e .      
Obtaining file:///Users/jason/Documents/GitHub/pymatgen
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
Requirement already satisfied: matplotlib>=3.8 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (3.9.1)
Requirement already satisfied: monty>=2024.5.24 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (2024.7.12)
Requirement already satisfied: networkx>=2.2 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (3.3)
Requirement already satisfied: palettable>=3.1.1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (3.3.3)
Requirement already satisfied: pandas>=2 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (2.2.2)
Requirement already satisfied: plotly>=4.5.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (5.22.0)
Requirement already satisfied: pybtex>=0.24.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (0.24.0)
Requirement already satisfied: requests>=2.32 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (2.32.3)
Requirement already satisfied: ruamel.yaml>=0.17.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (0.18.6)
Requirement already satisfied: scipy>=1.13.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (1.14.0)
Requirement already satisfied: spglib>=2.5.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (2.5.0)
Requirement already satisfied: sympy>=1.2 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (1.13.0)
Requirement already satisfied: tabulate>=0.9 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (0.9.0)
Requirement already satisfied: tqdm>=4.60 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (4.66.4)
Requirement already satisfied: uncertainties>=3.1.4 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (3.2.2)
Requirement already satisfied: joblib>=1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (1.4.2)
Requirement already satisfied: numpy>=1.25.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (1.26.4)
Requirement already satisfied: contourpy>=1.0.1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (1.2.1)
Requirement already satisfied: cycler>=0.10 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (0.12.1)
Requirement already satisfied: fonttools>=4.22.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (4.53.1)
Requirement already satisfied: kiwisolver>=1.3.1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (1.4.5)
Requirement already satisfied: packaging>=20.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (24.1)
Requirement already satisfied: pillow>=8 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (10.4.0)
Requirement already satisfied: pyparsing>=2.3.1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (3.1.2)
Requirement already satisfied: python-dateutil>=2.7 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (2.9.0.post0)
Requirement already satisfied: pytz>=2020.1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pandas>=2->pymatgen==2024.6.10) (2024.1)
Requirement already satisfied: tzdata>=2022.7 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pandas>=2->pymatgen==2024.6.10) (2024.1)
Requirement already satisfied: tenacity>=6.2.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from plotly>=4.5.0->pymatgen==2024.6.10) (8.5.0)
Requirement already satisfied: PyYAML>=3.01 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pybtex>=0.24.0->pymatgen==2024.6.10) (6.0.1)
Requirement already satisfied: latexcodec>=1.0.4 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pybtex>=0.24.0->pymatgen==2024.6.10) (3.0.0)
Requirement already satisfied: six in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pybtex>=0.24.0->pymatgen==2024.6.10) (1.16.0)
Requirement already satisfied: charset-normalizer<4,>=2 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from requests>=2.32->pymatgen==2024.6.10) (3.3.2)
Requirement already satisfied: idna<4,>=2.5 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from requests>=2.32->pymatgen==2024.6.10) (3.7)
Requirement already satisfied: urllib3<3,>=1.21.1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from requests>=2.32->pymatgen==2024.6.10) (2.2.2)
Requirement already satisfied: certifi>=2017.4.17 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from requests>=2.32->pymatgen==2024.6.10) (2024.7.4)
Requirement already satisfied: ruamel.yaml.clib>=0.2.7 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from ruamel.yaml>=0.17.0->pymatgen==2024.6.10) (0.2.8)
Requirement already satisfied: mpmath<1.4,>=1.1.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from sympy>=1.2->pymatgen==2024.6.10) (1.3.0)
Building wheels for collected packages: pymatgen
  Building editable for pymatgen (pyproject.toml) ... done
  Created wheel for pymatgen: filename=pymatgen-2024.6.10-0.editable-cp311-cp311-macosx_11_0_arm64.whl size=7492 sha256=2548cfbf6576818aa421af52eb35c4419af329a6ddc25491aefea55384cdd940
  Stored in directory: /private/var/folders/km/tznxn_d162bgzpqpzp9wcjn00000gn/T/pip-ephem-wheel-cache-w9pptfr8/wheels/86/ad/75/5dd22e6df08f3c8b6e3dfe775999afd0464ff20204e3661ff7
Successfully built pymatgen
Installing collected packages: pymatgen
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
emmet-core 0.84.1 requires pymatgen==2024.4.13, but you have pymatgen 2024.6.10 which is incompatible.
Successfully installed pymatgen-2024.6.10```

**Comment by jinlhr542** [comment]
> ```
> (3.11) jason@YaoshuXiedeMacBook-Pro pymatgen % pip install -e .      
> Obtaining file:///Users/jason/Documents/GitHub/pymatgen
>   Installing build dependencies ... done
>   Checking if build backend supports build_editable ... done
>   Getting requirements to build editable ... done
>   Preparing editable metadata (pyproject.toml) ... done
> Requirement already satisfied: matplotlib>=3.8 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (3.9.1)
> Requirement already satisfied: monty>=2024.5.24 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (2024.7.12)
> Requirement already satisfied: networkx>=2.2 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (3.3)
> Requirement already satisfied: palettable>=3.1.1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (3.3.3)
> Requirement already satisfied: pandas>=2 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (2.2.2)
> Requirement already satisfied: plotly>=4.5.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (5.22.0)
> Requirement already satisfied: pybtex>=0.24.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (0.24.0)
> Requirement already satisfied: requests>=2.32 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (2.32.3)
> Requirement already satisfied: ruamel.yaml>=0.17.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (0.18.6)
> Requirement already satisfied: scipy>=1.13.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (1.14.0)
> Requirement already satisfied: spglib>=2.5.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (2.5.0)
> Requirement already satisfied: sympy>=1.2 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (1.13.0)
> Requirement already satisfied: tabulate>=0.9 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (0.9.0)
> Requirement already satisfied: tqdm>=4.60 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (4.66.4)
> Requirement already satisfied: uncertainties>=3.1.4 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (3.2.2)
> Requirement already satisfied: joblib>=1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (1.4.2)
> Requirement already satisfied: numpy>=1.25.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pymatgen==2024.6.10) (1.26.4)
> Requirement already satisfied: contourpy>=1.0.1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (1.2.1)
> Requirement already satisfied: cycler>=0.10 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (0.12.1)
> Requirement already satisfied: fonttools>=4.22.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (4.53.1)
> Requirement already satisfied: kiwisolver>=1.3.1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (1.4.5)
> Requirement already satisfied: packaging>=20.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (24.1)
> Requirement already satisfied: pillow>=8 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (10.4.0)
> Requirement already satisfied: pyparsing>=2.3.1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (3.1.2)
> Requirement already satisfied: python-dateutil>=2.7 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from matplotlib>=3.8->pymatgen==2024.6.10) (2.9.0.post0)
> Requirement already satisfied: pytz>=2020.1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pandas>=2->pymatgen==2024.6.10) (2024.1)
> Requirement already satisfied: tzdata>=2022.7 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pandas>=2->pymatgen==2024.6.10) (2024.1)
> Requirement already satisfied: tenacity>=6.2.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from plotly>=4.5.0->pymatgen==2024.6.10) (8.5.0)
> Requirement already satisfied: PyYAML>=3.01 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pybtex>=0.24.0->pymatgen==2024.6.10) (6.0.1)
> Requirement already satisfied: latexcodec>=1.0.4 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pybtex>=0.24.0->pymatgen==2024.6.10) (3.0.0)
> Requirement already satisfied: six in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from pybtex>=0.24.0->pymatgen==2024.6.10) (1.16.0)
> Requirement already satisfied: charset-normalizer<4,>=2 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from requests>=2.32->pymatgen==2024.6.10) (3.3.2)
> Requirement already satisfied: idna<4,>=2.5 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from requests>=2.32->pymatgen==2024.6.10) (3.7)
> Requirement already satisfied: urllib3<3,>=1.21.1 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from requests>=2.32->pymatgen==2024.6.10) (2.2.2)
> Requirement already satisfied: certifi>=2017.4.17 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from requests>=2.32->pymatgen==2024.6.10) (2024.7.4)
> Requirement already satisfied: ruamel.yaml.clib>=0.2.7 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from ruamel.yaml>=0.17.0->pymatgen==2024.6.10) (0.2.8)
> Requirement already satisfied: mpmath<1.4,>=1.1.0 in /opt/anaconda3/envs/3.11/lib/python3.11/site-packages (from sympy>=1.2->pymatgen==2024.6.10) (1.3.0)
> Building wheels for collected packages: pymatgen
>   Building editable for pymatgen (pyproject.toml) ... done
>   Created wheel for pymatgen: filename=pymatgen-2024.6.10-0.editable-cp311-cp311-macosx_11_0_arm64.whl size=7492 sha256=2548cfbf6576818aa421af52eb35c4419af329a6ddc25491aefea55384cdd940
>   Stored in directory: /private/var/folders/km/tznxn_d162bgzpqpzp9wcjn00000gn/T/pip-ephem-wheel-cache-w9pptfr8/wheels/86/ad/75/5dd22e6df08f3c8b6e3dfe775999afd0464ff20204e3661ff7
> Successfully built pymatgen
> Installing collected packages: pymatgen
> ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
> emmet-core 0.84.1 requires pymatgen==2024.4.13, but you have pymatgen 2024.6.10 which is incompatible.
> Successfully installed pymatgen-2024.6.10```
> ```

Indeed there is an issue of the emmet-core dependency problem but it finally says successfully installed...

**Comment by DanielYang59** [comment]
`emmet-core` explicitly required `pymatgen==2024.4.13` https://github.com/materialsproject/emmet/issues/1020. My suggestion would be creating a separate Python env for developing `pymatgen` (or any other dedicated task) to avoid such version conflicts.

I believe it's better not to trust an incompletely resolved package dependency?

**Comment by jinlhr542** [comment]
yes, I found the problem. I am using mp-api which requires emmet-core, and emmet-core requires lower version of pymatgen... In this way I can only use the old version pymatgen now... Many thanks for your help there

**Review by janosh — CHANGES_REQUESTED** [review]
@jinlhr542 thanks for these improvements! could you add unit tests for the keywords you added?

**Comment by jinlhr542** [comment]
`import unittest
from pymatgen.core.structure import Structure
from pymatgen.core.lattice import Lattice
from pymatgen.analysis.interfaces import CoherentInterfaceBuilder, SubstrateAnalyzer

class TestCoherentInterfaceBuilder(unittest.TestCase):

    def setUp(self):
        #build substrate & film structure
        basis = [[0, 0, 0], [0.25, 0.25, 0.25]]
        self.substrate = Structure(
            Lattice.cubic(a=5.431),
            ["Si", "Si"],
            basis)
        self.film = substrate = Structure(
            Lattice.cubic(a=5.658),
            ["Ge", "Ge"],
            basis)

    def test_termination_searching(self):
        sub_analyzer = SubstrateAnalyzer()
        matches = list(sub_analyzer.calculate(substrate = self.substrate, film = self.film))
        cib = CoherentInterfaceBuilder(film_structure = self.film,
                               substrate_structure=self.substrate,
                               film_miller=matches[0].film_miller,
                               substrate_miller=matches[0].substrate_miller,
                               zslgen=sub_analyzer,termination_ftol=1e-4,label_index=True,\
                               filting_out_sym_slabs=False)
        self.assertTrue(cib.terminations == [('1_Ge_P4/mmm_1', '1_Si_P4/mmm_1'),\
                                             ('1_Ge_P4/mmm_1', '2_Si_P4/mmm_1'),\
                                             ('2_Ge_P4/mmm_1', '1_Si_P4/mmm_1'),\
                                             ('2_Ge_P4/mmm_1', '2_Si_P4/mmm_1')], \
"""
termination results wrong; the correct list should be:
[('1_Ge_P4/mmm_1', '1_Si_P4/mmm_1'),\
('1_Ge_P4/mmm_1', '2_Si_P4/mmm_1'),\
('2_Ge_P4/mmm_1', '1_Si_P4/mmm_1'),\
('2_Ge_P4/mmm_1', '2_Si_P4/mmm_1')].
""")
    # Add additional tests here to cover more functionality or edge cases as necessary

if __name__ == "__main__":
    unittest.main()`

**Comment by jinlhr542** [comment]
> @jinlhr542 thanks for these improvements! could you add unit tests for the keywords you added?

```
import unittest
from pymatgen.core.structure import Structure
from pymatgen.core.lattice import Lattice
from pymatgen.analysis.interfaces import CoherentInterfaceBuilder, SubstrateAnalyzer

class TestCoherentInterfaceBuilder(unittest.TestCase):

    def setUp(self):
        #build substrate & film structure
        basis = [[0, 0, 0], [0.25, 0.25, 0.25]]
        self.substrate = Structure(
            Lattice.cubic(a=5.431),
            ["Si", "Si"],
            basis)
        self.film = substrate = Structure(
            Lattice.cubic(a=5.658),
            ["Ge", "Ge"],
            basis)

    def test_termination_searching(self):
        sub_analyzer = SubstrateAnalyzer()
        matches = list(sub_analyzer.calculate(substrate = self.substrate, film = self.film))
        cib = CoherentInterfaceBuilder(film_structure = self.film,
                               substrate_structure=self.substrate,
                               film_miller=matches[0].film_miller,
                               substrate_miller=matches[0].substrate_miller,
                               zslgen=sub_analyzer,termination_ftol=1e-4,label_index=True,\
                               filting_out_sym_slabs=False)
        self.assertTrue(cib.terminations == [('1_Ge_P4/mmm_1', '1_Si_P4/mmm_1'),\
                                             ('1_Ge_P4/mmm_1', '2_Si_P4/mmm_1'),\
                                             ('2_Ge_P4/mmm_1', '1_Si_P4/mmm_1'),\
                                             ('2_Ge_P4/mmm_1', '2_Si_P4/mmm_1')], \
"""
termination results wrong; the correct list should be:
[('1_Ge_P4/mmm_1', '1_Si_P4/mmm_1'),\
('1_Ge_P4/mmm_1', '2_Si_P4/mmm_1'),\
('2_Ge_P4/mmm_1', '1_Si_P4/mmm_1'),\
('2_Ge_P4/mmm_1', '2_Si_P4/mmm_1')].
""")
    # Add additional tests here to cover more functionality or edge cases as necessary

if __name__ == "__main__":
    unittest.main()

```
Please use this one

**Comment by jinlhr542** [comment]
> @jinlhr542 thanks for these improvements! could you add unit tests for the keywords you added?

Could you please try this test?
```python
import unittest
from pymatgen.core.structure import Structure
from pymatgen.core.lattice import Lattice
from pymatgen.analysis.interfaces import CoherentInterfaceBuilder, SubstrateAnalyzer

class TestCoherentInterfaceBuilder(unittest.TestCase):

    def setUp(self):
        #build substrate & film structure
        basis = [[0, 0, 0], [0.25, 0.25, 0.25]]
        self.substrate = Structure(
            Lattice.cubic(a=5.431),
            ["Si", "Si"],
            basis)
        self.film = substrate = Structure(
            Lattice.cubic(a=5.658),
            ["Ge", "Ge"],
            basis)

    def test_termination_searching(self):
        sub_analyzer = SubstrateAnalyzer()
        matches = list(sub_analyzer.calculate(substrate = self.substrate, film = self.film))
        cib = CoherentInterfaceBuilder(film_structure = self.film,
                               substrate_structure=self.substrate,
                               film_miller=matches[0].film_miller,
                               substrate_miller=matches[0].substrate_miller,
                               zslgen=sub_analyzer,termination_ftol=1e-4,label_index=True,\
                               filting_out_sym_slabs=False)
        self.assertTrue(cib.terminations == [('1_Ge_P4/mmm_1', '1_Si_P4/mmm_1'),\
                                             ('1_Ge_P4/mmm_1', '2_Si_P4/mmm_1'),\
                                             ('2_Ge_P4/mmm_1', '1_Si_P4/mmm_1'),\
                                             ('2_Ge_P4/mmm_1', '2_Si_P4/mmm_1')], \
"""
termination results wrong; the correct list should be:
[('1_Ge_P4/mmm_1', '1_Si_P4/mmm_1'),
('1_Ge_P4/mmm_1', '2_Si_P4/mmm_1'),
('2_Ge_P4/mmm_1', '1_Si_P4/mmm_1'),
('2_Ge_P4/mmm_1', '2_Si_P4/mmm_1')].
""")

if __name__ == "__main__":
    unittest.main()
```

**Comment by DanielYang59** [comment]
Can you please insert that new test into `pymatgen/tests/analysis/interfaces/test_coherent_interface.py` ([usually the location of the unit tests follow the same structure of the module/package you wish to test](https://docs.pytest.org/en/stable/explanation/goodpractices.html)) and commit it? Such that it would be run automatically in the future.

Can you also do me a favor and rename the following unit test to `TestMillerIndexFinder`? Thanks:
https://github.com/materialsproject/pymatgen/blob/252efa7839e337c78b8633d32a3961aea949f58a/tests/core/test_surface.py#L693

**Comment by jinlhr542** [comment]
> TestMillerIndexFinder

Done

File: tests/core/test_surface.py [hunk 1]
```diff
@@ -846,3 +846,41 @@ def test_miller_index_from_sites(self):
         s3 = np.array([1.1595, 0.66943764, 0.9065])
         hkl = miller_index_from_sites(matrix, [s1, s2, s3])
         assert hkl == (2, -1, 0)
+
+class TestCoherentInterfaceBuilder(unittest.TestCase):
```
Comment (DanielYang59) [inline tests/core/test_surface.py]: Thanks for adding these test, but perhaps they should go into `pymatgen/tests/analysis/interfaces/test_coherent_interface.py`

Comment (jinlhr542) [reply tests/core/test_surface.py]: OK

Comment (DanielYang59) [reply tests/core/test_surface.py]: Thanks a ton for all those changes!

**Comment by jinlhr542** [comment]
> @jinlhr542 thanks for these improvements! could you add unit tests for the keywords you added?

Dear Janosh

Could you please check the commited tests? It will be appreciated if this pull request can be accepted soon.

File: src/pymatgen/core/surface.py [hunk 2]
```diff
@@ -1194,6 +1194,7 @@ def get_slabs(
         symmetrize: bool = False,
         repair: bool = False,
         ztol: float = 0,
+        filting_out_sym_slabs=True,
```
Comment (janosh) [inline src/pymatgen/core/surface.py:1197]:
```suggestion
        filter_out_sym_slabs: bool = True,
```

File: src/pymatgen/core/interface.py [hunk 3]
```diff
@@ -2838,8 +2838,14 @@ def from_slabs(
         return iface
 
 
-def label_termination(slab: Structure) -> str:
-    """Label the slab surface termination."""
+def label_termination(slab: Structure, ftol: float = 0.25, t_index: int | None = None) -> str:
```
Comment (janosh) [inline src/pymatgen/core/interface.py:2841]: how about just `tol` instead of `ftol`? does the `f` have any meaning? also, how about `t_index`->`t_idx` for brevity?

```suggestion
def label_termination(slab: Structure, ftol: float = 0.25, t_index: int | None = None) -> str:
```

**Review by janosh — CHANGES_REQUESTED** [review]
minor variable rename suggestions

**Comment by jinlhr542** [comment]
> minor variable rename suggestions



> minor variable rename suggestions

Modified accordingly, please review again.

**Review by janosh — APPROVED** [review]

File: src/pymatgen/analysis/interfaces/coherent_interfaces.py [hunk 4]
```diff
@@ -42,14 +45,21 @@ def __init__(
             film_miller: miller index of the film layer
             substrate_miller: miller index for the substrate layer
             zslgen: BiDirectionalZSL if you want custom lattice matching tolerances for coherency.
+            termination_ftol: tolerance to distinguish different terminating atomic planes.
+            label_index: whether to add an extra index at the beginning of the termination label.
+            filter_out_sym_slabs: whether to filter out identical slabs with different termination,
+            this might need to be set as False to find more non-identical terminations because slab
+            identity separately dose not mean combinational identity.
         """
```
[the remark this answers is not in the document]
Comment (jinlhr542) [reply src/pymatgen/analysis/interfaces/coherent_interfaces.py:53]: Yes. Specifically, for two combinations (A1, B1) and (A2, B2). 'A1 equal to A2' and 'B1 equal to B2' does not always make (A1, B2) equal to (A2, B2). When making interfaces the their symmetry operations need to be compatible to make a 'combinational' identity.

**Review by janosh — APPROVED** [review]
thanks a lot @jinlhr542! 👍

**Comment by DanielYang59** [comment]
@jinlhr542 Great work!
